electrum/lnrouter.py

```diff
@@ -161,12 +167,251 @@ def is_fee_sane(fee_msat: int, *, payment_amount_msat: int) -> bool:
+        # if we can't send over the channel, we should be able to send in the
+        # reverse direction
+        self.can_send_backward = amount
+
+    @property
+    def cannot_send_backward(self):
+        return self._cannot_send_backward
+
+    @cannot_send_backward.setter
+    def cannot_send_backward(self, amount):
+        if self._cannot_send_backward and self._cannot_send_backward < amount:
+            return
+        self._cannot_send_backward = amount
+        if self._can_send_backward and self._can_send_backward > self._cannot_send_backward:
+            self._can_send_backward = None
+        self.can_send_forward = amount
+
+    def can_send(self, is_forward_direction: bool):
+        # make info invalid after some time?
+        if is_forward_direction:
+            return self.can_send_forward
+        else:
+            return self.can_send_backward
+
+    def cannot_send(self, is_forward_direction: bool):
+        # make info invalid after some time?
+        if is_forward_direction:
+            return self.cannot_send_forward
+        else:
+            return self.cannot_send_backward
+
+    def update_can_send(self, is_forward_direction: bool, amount: int):
+        if is_forward_direction:
+            self.can_send_forward = amount
+        else:
+            self.can_send_backward = amount
+
+    def update_cannot_send(self, is_forward_direction: bool, amount: int):
+        if is_forward_direction:
+            self.cannot_send_forward = amount
+        else:
+            self.cannot_send_backward = amount
+
+    def __repr__(self):
+        return f"forward: can send: {self._can_send_forward}, cannot send: {self._cannot_send_forward}, \n" \
+               f"backward: can send: {self._can_send_backward} cannot send: {self._cannot_send_backward}, \n" \
+               f"blacklisted: {self.blacklisted}"
+
+
+class LiquidityHints:
+    """Implements liquidity hints for channels in the graph.
+
+    This class can be used to update liquidity information about channels in the
+    graph. Implements a penalty function for edge weighting in the pathfinding
+    algorithm that favors channels which can route payments and penalizes
+    channels that cannot.
+    """
+    # TODO: incorporate in-flight htlcs
+    # TODO: use timestamps for can/not_send to make them None after some time?
+    # TODO: hints based on node pairs only (shadow channels, non-strict forwarding)?
+    def __init__(self):
+        self.lock = RLock()
+        self._liquidity_hints: Dict[Tuple[bytes, bytes, ShortChannelID], LiquidityHint] = {}
+
+    @staticmethod
+    def _hint_key(node1: bytes, node2: bytes, channel_id: ShortChannelID):
+        return (node1, node2, channel_id) if node1 < node2 else (node2, node1, channel_id)
+
+    @with_lock
+    def get_hint(self, node1: bytes, node2: bytes, channel_id: ShortChannelID):
+        key = self._hint_key(node1, node2, channel_id)
+        return self._liquidity_hints.get(key, LiquidityHint())
```

Comment:
There is a potential API issue here: maybe the default value of the get should be saved in the map before returning it.
If this method creates a new `LiquidityHint()` and the caller starts mutating it, the caller might expect the mutations not to get lost.
If you saved the newly created `LiquidityHint()` in the map before returning it, you could use `get_hint()` in `update_can_send`, `update_cannot_send`, and `blacklist`.

Reply:
Great point, I applied your suggestions. Only in the penalty function I'm using the dict get, as otherwise many insignificant hints would be created.